File: crates/stateset-embedded/src/vector.rs
//! Vector search operations for semantic similarity search
//!
//! This module provides hybrid semantic + lexical search using embeddings
//! and BM25 (SQLite FTS5) for improved relevance.
//! If FTS5 isn't available, it transparently falls back to embedding-only search.
//!
//! # Example
//!
//! ```rust,ignore
//! use stateset_embedded::Commerce;
//!
//! let commerce = Commerce::new("./store.db")?;
//!
//! // Initialize vector search with OpenAI API key
//! let vector = commerce.vector(std::env::var("OPENAI_API_KEY")?);
//!
//! // Index a product for search
//! vector.index_product(&product)?;
//!
//! // Search for similar products
//! let results = vector.search_products("wireless bluetooth headphones", 10)?;
//! for result in results {
//!     println!("{}: {} (score: {:.2})", result.entity.name, result.entity.id, result.score);
//! }
//! ```

use stateset_core::{
    Customer, EmbeddingService, EmbeddingStats, EntityType, InventoryItem, Order, Product, Result,
    VectorRepository, VectorSearchResult,
};
use stateset_db::sqlite::SqliteVectorRepository;
use std::sync::Arc;

/// Vector search operations
pub struct Vector {
    repo: SqliteVectorRepository,
    embedding_service: Arc<EmbeddingService>,
}

impl Vector {
    /// Create a new Vector instance
    pub(crate) fn new(repo: SqliteVectorRepository, api_key: String) -> Self {
        Self {
            repo,
            embedding_service: Arc::new(EmbeddingService::new(api_key)),
        }
    }

    /// Create with custom embedding model
    pub fn with_model(
        repo: SqliteVectorRepository,
        api_key: String,
        model: String,
        dimensions: usize,
    ) -> Self {
        Self {
            repo,
            embedding_service: Arc::new(EmbeddingService::with_model(api_key, model, dimensions)),
        }
    }

    // ========================================================================
    // Product Operations
    // ========================================================================

    /// Index a product for vector search
    pub fn index_product(&self, product: &Product) -> Result<()> {
        let text = EmbeddingService::product_text(product);
        let result = self.embedding_service.embed(&text)?;
        self.repo.store_embedding(
            EntityType::Product,
            &product.id.to_string(),
            &result.embedding,
            &result.text_hash,
            self.embedding_service.model(),
        )
    }

    /// Search for products similar to the query text
    pub fn search_products(
        &self,
        query: &str,
        limit: usize,
    ) -> Result<Vec<VectorSearchResult<Product>>> {
        let result = self.embedding_service.embed(query)?;
        self.repo
            .search_products_hybrid(&result.embedding, query, limit)
    }

    /// Search products using a pre-computed embedding
    pub fn search_products_by_embedding(
        &self,
        embedding: &[f32],
        limit: usize,
    ) -> Result<Vec<VectorSearchResult<Product>>> {
        self.repo.search_products(embedding, limit)
    }

    /// Remove a product from the vector index
    pub fn unindex_product(&self, product_id: &str) -> Result<()> {
        self.repo.delete_embedding(EntityType::Product, product_id)
    }

    /// Index multiple products in batch
    pub fn index_products(&self, products: &[Product]) -> Result<usize> {
        let mut indexed = 0;

        // Process in batches of 100 for API efficiency
        for chunk in products.chunks(100) {
            let texts: Vec<String> = chunk.iter().map(EmbeddingService::product_text).collect();

            let results = self.embedding_service.embed_batch(&texts)?;

            for (product, result) in chunk.iter().zip(results.iter()) {
                self.repo.store_embedding(
                    EntityType::Product,
                    &product.id.to_string(),
                    &result.embedding,
                    &result.text_hash,
                    self.embedding_service.model(),
                )?;
                indexed += 1;
            }
        }

        Ok(indexed)
    }

    // ========================================================================
    // Customer Operations
    // ========================================================================

    /// Index a customer for vector search
    pub fn index_customer(&self, customer: &Customer) -> Result<()> {
        let text = EmbeddingService::customer_text(customer);
        let result = self.embedding_service.embed(&text)?;
        self.repo.store_embedding(
            EntityType::Customer,
            &customer.id.to_string(),
            &result.embedding,
            &result.text_hash,
            self.embedding_service.model(),
        )
    }

    /// Search for customers similar to the query text
    pub fn search_customers(
        &self,
        query: &str,
        limit: usize,
    ) -> Result<Vec<VectorSearchResult<Customer>>> {
        let result = self.embedding_service.embed(query)?;
        self.repo
            .search_customers_hybrid(&result.embedding, query, limit)
    }

    /// Remove a customer from the vector index
    pub fn unindex_customer(&self, customer_id: &str) -> Result<()> {
        self.repo
            .delete_embedding(EntityType::Customer, customer_id)
    }

    /// Index multiple customers in batch
    pub fn index_customers(&self, customers: &[Customer]) -> Result<usize> {
        let mut indexed = 0;

        for chunk in customers.chunks(100) {
            let texts: Vec<String> = chunk.iter().map(EmbeddingService::customer_text).collect();

            let results = self.embedding_service.embed_batch(&texts)?;

            for (customer, result) in chunk.iter().zip(results.iter()) {
                self.repo.store_embedding(
                    EntityType::Customer,
                    &customer.id.to_string(),
                    &result.embedding,
                    &result.text_hash,
                    self.embedding_service.model(),
                )?;
                indexed += 1;
            }
        }

        Ok(indexed)
    }

    // ========================================================================
    // Order Operations
    // ========================================================================

    /// Index an order for vector search
    pub fn index_order(&self, order: &Order) -> Result<()> {
        let text = EmbeddingService::order_text(order);
        let result = self.embedding_service.embed(&text)?;
        self.repo.store_embedding(
            EntityType::Order,
            &order.id.to_string(),
            &result.embedding,
            &result.text_hash,
            self.embedding_service.model(),
        )
    }

    /// Search for orders similar to the query text
    pub fn search_orders(
        &self,
        query: &str,
        limit: usize,
    ) -> Result<Vec<VectorSearchResult<Order>>> {
        let result = self.embedding_service.embed(query)?;
        self.repo
            .search_orders_hybrid(&result.embedding, query, limit)
    }

    /// Remove an order from the vector index
    pub fn unindex_order(&self, order_id: &str) -> Result<()> {
        self.repo.delete_embedding(EntityType::Order, order_id)
    }

    /// Index multiple orders in batch
    pub fn index_orders(&self, orders: &[Order]) -> Result<usize> {
        let mut indexed = 0;

        for chunk in orders.chunks(100) {
            let texts: Vec<String> = chunk.iter().map(EmbeddingService::order_text).collect();

            let results = self.embedding_service.embed_batch(&texts)?;

            for (order, result) in chunk.iter().zip(results.iter()) {
                self.repo.store_embedding(
                    EntityType::Order,
                    &order.id.to_string(),
                    &result.embedding,
                    &result.text_hash,
                    self.embedding_service.model(),
                )?;
                indexed += 1;
            }
        }

        Ok(indexed)
    }

    // ========================================================================
    // Inventory Operations
    // ========================================================================

    /// Index an inventory item for vector search
    pub fn index_inventory_item(&self, item: &InventoryItem) -> Result<()> {
        let text = EmbeddingService::inventory_item_text(item);
        let result = self.embedding_service.embed(&text)?;
        self.repo.store_embedding(
            EntityType::InventoryItem,
            &item.id.to_string(),
            &result.embedding,
            &result.text_hash,
            self.embedding_service.model(),
        )
    }

    /// Search for inventory items similar to the query text
    pub fn search_inventory(
        &self,
        query: &str,
        limit: usize,
    ) -> Result<Vec<VectorSearchResult<InventoryItem>>> {
        let result = self.embedding_service.embed(query)?;
        self.repo
            .search_inventory_hybrid(&result.embedding, query, limit)
    }

    /// Remove an inventory item from the vector index
    pub fn unindex_inventory_item(&self, item_id: &str) -> Result<()> {
        self.repo
            .delete_embedding(EntityType::InventoryItem, item_id)
    }

    /// Index multiple inventory items in batch
    pub fn index_inventory_items(&self, items: &[InventoryItem]) -> Result<usize> {
        let mut indexed = 0;

        for chunk in items.chunks(100) {
            let texts: Vec<String> = chunk
                .iter()
                .map(EmbeddingService::inventory_item_text)
                .collect();

            let results = self.embedding_service.embed_batch(&texts)?;

            for (item, result) in chunk.iter().zip(results.iter()) {
                self.repo.store_embedding(
                    EntityType::InventoryItem,
                    &item.id.to_string(),
                    &result.embedding,
                    &result.text_hash,
                    self.embedding_service.model(),
                )?;
                indexed += 1;
            }
        }

        Ok(indexed)
    }

    // ========================================================================
    // Utility Operations
    // ========================================================================

    /// Get embedding statistics
    pub fn stats(&self) -> Result<EmbeddingStats> {
        self.repo.get_stats()
    }

    /// Check if an entity has been indexed
    pub fn is_indexed(&self, entity_type: EntityType, entity_id: &str) -> Result<bool> {
        self.repo.has_embedding(entity_type, entity_id)
    }

    /// Clear all embeddings for an entity type
    pub fn clear(&self, entity_type: EntityType) -> Result<u64> {
        self.repo.clear_embeddings(entity_type)
    }

    /// Clear all embeddings
    pub fn clear_all(&self) -> Result<u64> {
        let mut total = 0;
        total += self.repo.clear_embeddings(EntityType::Product)?;
        total += self.repo.clear_embeddings(EntityType::Customer)?;
        total += self.repo.clear_embeddings(EntityType::Order)?;
        total += self.repo.clear_embeddings(EntityType::InventoryItem)?;
        Ok(total)
    }

    /// Generate an embedding for text without storing it
    pub fn embed(&self, text: &str) -> Result<Vec<f32>> {
        let result = self.embedding_service.embed(text)?;
        Ok(result.embedding)
    }
}

File: cli/src/commands/vector.js
/**
 * Vector Commands Module
 */

function getVectorSearch(commerce) {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    throw new Error('OPENAI_API_KEY environment variable is required for vector search');
  }
  return commerce.vector(apiKey);
}

function parseLimit(value, usage) {
  if (value === undefined) return undefined;
  const parsed = Number.parseInt(value, 10);
  if (!Number.isInteger(parsed) || parsed <= 0) throw new Error(usage);
  return parsed;
}

export async function execute(action, args, { commerce, output, jsonOutput }) {
  const vector = getVectorSearch(commerce);

  switch (action) {
    case 'search-products': {
      const [query, limitRaw] = args;
      if (!query) throw new Error('Usage: vector search-products <query> [limit]');
      const results = await vector.searchProducts(
        query,
        parseLimit(limitRaw, 'Usage: vector search-products <query> [limit]') || 10,
      );
      return formatSearchResults(
        results.map((result) => ({
          id: result.product.id,
          name: result.product.name,
          description: result.product.description,
          score: result.score.toFixed(3),
        })),
        { output, jsonOutput },
      );
    }

    case 'search-customers': {
      const [query, limitRaw] = args;
      if (!query) throw new Error('Usage: vector search-customers <query> [limit]');
      const results = await vector.searchCustomers(
        query,
        parseLimit(limitRaw, 'Usage: vector search-customers <query> [limit]') || 10,
      );
      return formatSearchResults(
        results.map((result) => ({
          id: result.customer.id,
          name: `${result.customer.firstName} ${result.customer.lastName}`,
          description: result.customer.email,
          score: result.score.toFixed(3),
        })),
        { output, jsonOutput },
      );
    }

    case 'search-orders': {
      const [query, limitRaw] = args;
      if (!query) throw new Error('Usage: vector search-orders <query> [limit]');
      const results = await vector.searchOrders(
        query,
        parseLimit(limitRaw, 'Usage: vector search-orders <query> [limit]') || 10,
      );
      return formatSearchResults(
        results.map((result) => ({
          id: result.order.id,
          name: result.order.orderNumber,
          description: result.order.status,
          score: result.score.toFixed(3),
        })),
        { output, jsonOutput },
      );
    }

    case 'search-inventory': {
      const [query, limitRaw] = args;
      if (!query) throw new Error('Usage: vector search-inventory <query> [limit]');
      const results = await vector.searchInventory(
        query,
        parseLimit(limitRaw, 'Usage: vector search-inventory <query> [limit]') || 10,
      );
      return formatSearchResults(
        results.map((result) => ({
          id: result.item.id,
          name: result.item.name,
          description: result.item.sku,
          score: result.score.toFixed(3),
        })),
        { output, jsonOutput },
      );
    }

    case 'index-product': {
      const productId = args[0];
      if (!productId) throw new Error('Usage: vector index-product <productId>');
      await vector.indexProduct(productId);
      return { formatted: `Indexed product ${productId} for vector search` };
    }

    case 'index-customer': {
      const customerId = args[0];
      if (!customerId) throw new Error('Usage: vector index-customer <customerId>');
      await vector.indexCustomer(customerId);
      return { formatted: `Indexed customer ${customerId} for vector search` };
    }

    case 'index-order': {
      const orderId = args[0];
      if (!orderId) throw new Error('Usage: vector index-order <orderId>');
      await vector.indexOrder(orderId);
      return { formatted: `Indexed order ${orderId} for vector search` };
    }

    case 'index-inventory': {
      const itemId = args[0];
      if (!itemId) throw new Error('Usage: vector index-inventory <itemId>');
      await vector.indexInventoryItem(itemId);
      return { formatted: `Indexed inventory item ${itemId} for vector search` };
    }

    case 'stats': {
      const stats = await vector.stats();
      const result = {
        model: stats.model,
        dimensions: stats.dimensions,
        counts: {
          products: stats.productCount,
          customers: stats.customerCount,
          orders: stats.orderCount,
          inventory: stats.inventoryCount,
        },
      };
      return jsonOutput
        ? result
        : {
            result,
            formatted:
              `Vector stats\n` +
              `${'-'.repeat(22)}\n` +
              `Model:       ${result.model}\n` +
              `Dimensions:  ${result.dimensions}\n` +
              `Products:    ${result.counts.products}\n` +
              `Customers:   ${result.counts.customers}\n` +
              `Orders:      ${result.counts.orders}\n` +
              `Inventory:   ${result.counts.inventory}`,
          };
    }

    case 'clear': {
      const entityType = args[0];
      if (!entityType) throw new Error('Usage: vector clear <entityType>');
      const count = await vector.clear(entityType);
      return { formatted: `Cleared ${count} ${entityType} embeddings` };
    }

    case 'clear-all': {
      const count = await vector.clearAll();
      return { formatted: `Cleared ${count} embeddings` };
    }

    case 'reindex-all': {
      await vector.clearAll();
      const products = await vector.indexAllProducts();
      const customers = await vector.indexAllCustomers();
      const orders = await vector.indexAllOrders();
      const inventory = await vector.indexAllInventory();
      const result = {
        products,
        customers,
        orders,
        inventory,
        total: products + customers + orders + inventory,
      };
      return jsonOutput
        ? result
        : {
            result,
            formatted:
              `Vector reindex complete\n` +
              `${'-'.repeat(32)}\n` +
              `Products:    ${products}\n` +
              `Customers:   ${customers}\n` +
              `Orders:      ${orders}\n` +
              `Inventory:   ${inventory}\n` +
              `Total:       ${result.total}`,
          };
    }

    default:
      throw new Error(
        `Unknown action: vector ${action}\n\n` +
          'Available actions:\n' +
          '  search-products <query> [limit]    Search products\n' +
          '  search-customers <query> [limit]   Search customers\n' +
          '  search-orders <query> [limit]      Search orders\n' +
          '  search-inventory <query> [limit]   Search inventory\n' +
          '  index-product <productId>          Index product\n' +
          '  index-customer <customerId>        Index customer\n' +
          '  index-order <orderId>              Index order\n' +
          '  index-inventory <itemId>           Index inventory item\n' +
          '  stats                              Get vector stats\n' +
          '  clear <entityType>                 Clear embeddings for entity type\n' +
          '  clear-all                          Clear all embeddings\n' +
          '  reindex-all                        Rebuild all embeddings',
      );
  }
}

function formatSearchResults(results, { output, jsonOutput }) {
  if (jsonOutput) return results;
  if (results.length === 0) return { formatted: 'No vector search results found.' };
  const formatted = output.table(results, [
    { key: 'id', header: 'ID' },
    { key: 'name', header: 'Name' },
    { key: 'description', header: 'Description' },
    { key: 'score', header: 'Score', align: 'right' },
  ]);
  return { results, formatted };
}

export const metadata = {
  name: 'vector',
  aliases: ['vec', 'semantic'],
  description: 'Vector search and indexing commands',
  actions: {
    'search-products': { description: 'Search products', args: ['<query>', '[limit]'] },
    'search-customers': { description: 'Search customers', args: ['<query>', '[limit]'] },
    'search-orders': { description: 'Search orders', args: ['<query>', '[limit]'] },
    'search-inventory': { description: 'Search inventory', args: ['<query>', '[limit]'] },
    'index-product': { description: 'Index product', args: ['<productId>'] },
    'index-customer': { description: 'Index customer', args: ['<customerId>'] },
    'index-order': { description: 'Index order', args: ['<orderId>'] },
    'index-inventory': { description: 'Index inventory item', args: ['<itemId>'] },
    stats: { description: 'Get vector stats', args: [] },
    clear: { description: 'Clear embeddings for entity type', args: ['<entityType>'] },
    'clear-all': { description: 'Clear all embeddings', args: [] },
    'reindex-all': { description: 'Rebuild all embeddings', args: [] },
  },
};

export default { execute, metadata };
